core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -458,10 +461,12 @@ private void cleanUncommittedAppends(Set<ManifestFile> committed) {
       this.cachedNewManifest = null;
     }
 
-    if (cachedNewDeleteManifest != null && !committed.contains(cachedNewDeleteManifest)) {
-      deleteFile(cachedNewDeleteManifest.path());
-      this.cachedNewDeleteManifest = null;
+    for (ManifestFile cachedNewDeleteManifest : cachedNewDeleteManifests) {
+      if (!committed.contains(cachedNewDeleteManifest)) {
+        deleteFile(cachedNewDeleteManifest.path());
+      }
     }
+    this.cachedNewDeleteManifests.clear();
```

Comment:
Do we need the explicit clear here? Are we just trying to free it up for GC early?

Reply:
This logic seems a little different than it was previously? 

Before
```
if committed doesn't contain cachedNewDeleteManifest
  deleteFile()
  clear cachedNewDeleteManifest
```

```
for any cachedNewDeleteManifest
  if commited doesn't contain cachedNewDeleteManifest
     deleteFile
clear all cachedDeleteManifests
```
I'm still trying to understand the check here but it seems like we will clear out all manifests even if some of them are committed?

Reply:
You are right, I'll update this place.

Reply:
Updated to use `LinkedList` and `listIterator`.